src/rest-api/class-base-api-controller.php

```diff
@@ -0,0 +1,149 @@
+<?php
+/**
+ * Base API Controller
+ *
+ * @package Parsely
+ * @since   3.17.0
+ */
+
+declare(strict_types=1);
+
+namespace Parsely\REST_API;
+
+use Parsely\Parsely;
+
+/**
+ * Base API Controller.
+ *
+ * Used to define the namespace, version, and endpoints for an API controller. API controllers
+ * should extend this class and implement the `init` method to register endpoints.
+ *
+ * @since 3.17.0
+ */
+abstract class Base_API_Controller {
+
+	/**
+	 * The API namespace.
+	 *
+	 * @var string|false
+	 */
+	public const NAMESPACE = false;
+
+	/**
+	 * The API version.
+	 *
+	 * @var string
+	 */
+	public const VERSION = '';
+
+	/**
+	 * The route prefix, which acts as a namespace for the endpoints.
+	 *
+	 * @since 3.17.0
+	 *
+	 * @var string
+	 */
+	public const ROUTE_PREFIX = '';
+
+	/**
+	 * The endpoints.
+	 *
+	 * @var Base_Endpoint[]
+	 */
+	public $endpoints;
+
+	/**
+	 * The Parsely instance.
+	 *
+	 * @var Parsely
+	 */
+	public $parsely;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Parsely $parsely The Parsely instance.
+	 * @throws \UnexpectedValueException If the namespace is not defined.
+	 */
+	public function __construct( Parsely $parsely ) {
+		$this->parsely   = $parsely;
+		$this->endpoints = array();
+
+		if ( static::NAMESPACE === false ) {
+			throw new \UnexpectedValueException( 'The API controller must define a namespace.' );
+		}
+	}
+
+	/**
+	 * Initialize the API controller.
```

Comment:
Some functions in this file need a third person summary:

```suggestion
	 * Initializes the API controller.
```

It's worth checking other files for this kind of thing as well.

This specific comment also mises a `@since` tag.